deviceplugin/usb.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2023 the generic-device-plugin authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deviceplugin
+
+import (
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const (
+	usbDevicesDir              = "/sys/bus/usb/devices/"
+	usbDevicesDirVendorIDFile  = "idVendor"
+	usbDevicesDirProductIDFile = "idProduct"
+	usbDevicesDirBusFile       = "busnum"
+	usbDevicesDirBusDevFile    = "devnum"
+	usbDevBus                  = "/dev/bus/usb/%+04d/%+04d"
+)
+
+// USBSpec represents a USB device specification that should be discovered.
+// A USB device must match exactly on all the given attributes to pass.
+type USBSpec struct {
+	// Vendor is the USB Vendor ID of the device to match on.
+	// (Both of these get mangled to uint16 for processing - but you should use the hexadecimal representation.)
+	Vendor usbID `json:"vendor"`
```

Comment:
Ah, one thing that just occurred to me is that if `USBSpec` is exported, then the _types_ of all of its exported fields should be exported as well. This is primarily is just to be a good citizen for anything that becomes part of the public API of a go package so that if anyone is interacting with this struct, that they can interact with it freely 
```suggestion
	Vendor USBID `json:"vendor"`
```

Reply:
Implemented in 61c93f6, though I'm starting to get the feeling that breaking out into a go-native USB handling library is becoming more and more likely...